src/Idris/ProcessIdr.idr

```diff
@@ -223,26 +223,45 @@ gc = primIO $ prim__gc 4
 export
 addPublicHash : {auto c : Ref Ctxt Defs} ->
                 (Bool, (Namespace, Int)) -> Core ()
-addPublicHash (True, (mod, h)) = do addHash mod; addHash h
+addPublicHash (True, (mod, h)) = do addHash mod
+                                    addHash h
+                                    log "module.hash" 15 "Adding hash for a public import of \{show mod}"
 addPublicHash _ = pure ()
 
-||| If the source file is older
-unchangedTime : (sourceFileName : String) -> (ttcFileName : String) -> Core Bool
-unchangedTime sourceFileName ttcFileName
-  = do srcTime <- modTime sourceFileName
-       ttcTime <- modTime ttcFileName
-       pure $ srcTime <= ttcTime
-
-
-||| If the source file hash hasn't changed
-unchangedHash : (hashFn : Maybe String) -> (sourceFileName : String) -> (ttcFileName : String) -> Core Bool
-unchangedHash hashFn sourceFileName ttcFileName
+||| Determine if the TTC is outdated based on any of the given
+||| source or dependency source file names.
+export
+isTTCOutdated : {auto c : Ref Ctxt Defs} ->
+                (ttcFile : String) ->
+                (sourceFiles : List String) ->
+                Core Bool
+isTTCOutdated ttcFile sourceFiles
+  = do ttcTime  <- modTime ttcFile
+       srcTimes <- traverse modTime sourceFiles
+       log "module.hash" 20 $
+         unwords $
+           [ "Checking whether source code mod times are newer than \{show ttcTime};"
+           , "src times:"
+           , show srcTimes
+           ]
+       pure $ any (>= ttcTime) srcTimes
+
+||| If the source files hash hasn't changed
+export
+unchangedHash : (hashFn : Maybe String) -> (ttcFileName : String) -> (sourceFileName : String) -> Core Bool
+unchangedHash hashFn ttcFileName sourceFileName
   = do Just sourceCodeHash        <- hashFileWith hashFn sourceFileName
              | _ => pure False
        (Just storedSourceHash, _) <- readHashes ttcFileName
              | _ => pure False
        pure $ sourceCodeHash == storedSourceHash
 
+||| Determine if any of the given source files have changed.
+export
+unchangedHashes : (hashFn : Maybe String) -> (ttcFileName : String) -> (sourceFileNames : List String) -> Core Bool
+unchangedHashes hashFn ttcFileName
+  = map (all id) . traverse (unchangedHash hashFn ttcFileName)
```

Comment:
Does this mean we keep re-reading the hash for `ttcFileName`? Is that expensive?

Reply:
This code does re-read the hash for each `sourceFileName`, but I think that optimizing that away here would be premature given that this function is only called with a single item in `sourceFileNames` in practice.

Now that it's been a bit and I am looking back, I don't like that I chose to create this function to match the signature of the `isTTCOutdated` function just for the sake of the call site below. I am going to push a change that entirely deletes this function and instead reworks the call site slightly.